pkg/agent/datapath/natBridge.go

```diff
@@ -216,13 +273,233 @@ func (n *NatBridge) DelLBIP(svcID, ip string) error {
+			return err
+		}
+		svcOvsCache.SetLBFlow(ip, curEntry.PortName, nil)
+	}
+
+	sessionAffinityFlows := svcOvsCache.GetSessionAffinityFlowsByIP(ip)
+	for i := range sessionAffinityFlows {
+		curEntry := sessionAffinityFlows[i]
+		if curEntry.Flow == nil {
+			continue
+		}
+		if err := curEntry.Flow.Delete(); err != nil {
+			log.Errorf("Delete session affinity flow failed for lb ip: %s, port: %s, svcID: %s, flow: %+v, err: %s", ip, curEntry.PortName, svcID, curEntry.Flow, err)
+			return err
+		}
+		svcOvsCache.SetSessionAffinityFlow(ip, curEntry.PortName, nil)
+	}
+
+	log.Infof("Dp success DelLBIP for service %s ip %s", svcID, ip)
+	return nil
+}
+
+func (n *NatBridge) AddLBPort(svcID string, port *proxycache.Port, ips []string, sessionAffinityTimeout int32) error {
+	if port == nil {
+		return nil
+	}
+
+	portName := port.Name
+	svcOvsCache := n.svcIndexCache.GetSvcOvsInfoAndInitIfEmpty(svcID)
+
+	var err error
+	gp := svcOvsCache.GetGroup(portName)
+	if gp == nil {
+		gp, err = n.createEmptyGroup()
+		if err != nil {
+			log.Errorf("Failed to create empty group for service %s portname %s, err: %s", svcID, portName, err)
+			return err
+		}
+		svcOvsCache.SetGroup(portName, gp)
+	}
+	gpID := gp.GroupID
+
+	for i := range ips {
+		ipDa := net.ParseIP(ips[i])
+		if ipDa == nil {
+			log.Errorf("Invalid lb ip %s for service %s", ips[i], svcID)
+			return fmt.Errorf("invalid lb ip: %s", ips[i])
+		}
+		if svcOvsCache.GetLBFlow(ips[i], portName) != nil {
+			log.Infof("The lb flow for service %s ip %s port %s has been installed, skip create it", svcID, ips[i], portName)
+			continue
+		}
+		lbFlow, err := n.addLBFlow(&ipDa, port.Protocol, port.Port, gpID)
+		if err != nil {
+			log.Errorf("Failed to add a lb flow for service %s, ip: %s, portname: %s, err: %s", svcID, ips[i], portName, err)
+			return err
+		}
+		svcOvsCache.SetLBFlow(ips[i], portName, lbFlow)
+	}
+
+	if sessionAffinityTimeout <= 0 {
+		log.Infof("Dp success AddLBPort for service %s port %s", svcID, portName)
+		return nil
+	}
+	for i := range ips {
+		ipDa := net.ParseIP(ips[i])
+		if ipDa == nil {
+			log.Errorf("Invalid lb ip %s for service %s", ips[i], svcID)
+			return fmt.Errorf("invalid lb ip: %s", ips[i])
+		}
+		if svcOvsCache.GetSessionAffinityFlow(ips[i], portName) != nil {
+			log.Infof("The session affinity flow for service %s ip %s port %s has been installed, skip create it", svcID, ips[i], portName)
+			continue
+		}
+		sessionFlow, err := n.addSessionAffinityFlow(ipDa, port.Protocol, port.Port, sessionAffinityTimeout)
+		if err != nil {
+			log.Errorf("Failed to add session affinity flow for service %s ip %s, port %s, err: %s", svcID, ips[i], portName, err)
 			return err
 		}
-		svcOvsCache.DelLBFlow(ip, lbFlows[i].PortName)
+		svcOvsCache.SetSessionAffinityFlow(ips[i], portName, sessionFlow)
 	}
-	// todo process sessionAffinity
+
+	log.Infof("Dp success AddLBPort for service %s port %s", svcID, portName)
```

Comment:
generally not export the func name in the log.

Reply:
done